### PR TITLE
simplify LRU for parsedClasses in the ByteCodeRepository

### PR DESCRIPTION
**Pull request by mkeskells**
should reduce CPU and memory usage when optimising

### CODE REVIEW

File: src/compiler/scala/tools/nsc/backend/jvm/opt/ByteCodeRepository.scala [hunk 1]
```diff
@@ -35,68 +35,33 @@ abstract class ByteCodeRepository {
...
   }
 
-  private object lruCounter extends AtomicLong(0l) with collection.generic.Clearable {
-    def clear(): Unit = { this.set(0l) }
-  }
-  recordPerRunCache(lruCounter)
-
-  /**
-   * Prevent the code repository from growing too large. Profiling reveals that the average size
-   * of a ClassNode is about 30 kb. I observed having 17k+ classes in the cache, i.e., 500 mb.
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/opt/ByteCodeRepository.scala:67]: could you move this comment up to the definition of `private val maxCacheSize`?

File: src/compiler/scala/tools/nsc/backend/jvm/opt/LruMap.scala [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  def apply[K,V](maxSize:Int, threadsafe:Boolean): Map[K,V] = {
+    require (maxSize > 0)
+    val basic = new LruMapImpl[K,V](maxSize)
+    val threaded = if (threadsafe) Collections.synchronizedMap(basic) else basic
+
+    threaded.asScala
+  }
+
+  private class LruMapImpl[K,V](maxSize: Int) extends LinkedHashMap[K,V] {
+    override def removeEldestEntry(eldest: JMap.Entry[K, V]): Boolean = {
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/opt/LruMap.scala:17]: TIL! Thanks.

**Review by lrytz — COMMENTED** [review]
Looks good otherwise.

**Comment by lrytz** [comment]
LGTM, please squash the changes into a single commit.

**Comment by mkeskells** [comment]
@lrytz will squash but there is an outstanding query for you here though - see "I dont see why this was concurrent, but retained the thread safety. @lrytz what are the required semantics"

**Comment by lrytz** [comment]
I guess it depends how far we go with parallelizing the backend. The bytecode repo is used by the inliner to get bytecode, but also to obtain ClassBTypes. For all classes referenced in the source code, CodeGen creates a ClassBType from the corresponding Symbol. However, for references to "new" classes introduced by inlining, we don't have a symbol and create the ClassBType from the classfile. I think this doesn't happen eagerly in the inliner, but can happen at a later point when traversing the bytecode, for example local optimizations (`eliminateRedundantCasts`) or collecting references to nested classes (`NestedClassesCollector`).

During classfile writing, ClassBTypes are also accessed to compute stack map frames (`ClassWriterWithBTypeLub`). However, at this point the implementation expects the ClassBTypes to already exist in the cache (it calls `cachedClassBType(internalName).get`).

**Comment by mkeskells** [comment]
@lrytz OK - will add a comment that concurrency is an enabler for future work

File: src/compiler/scala/tools/nsc/backend/jvm/opt/ByteCodeRepository.scala (hunk 1 above)
Comment (mkeskells) [reply src/compiler/scala/tools/nsc/backend/jvm/opt/ByteCodeRepository.scala:67]: done in e28ad86

**Comment by mkeskells** [comment]
@lrytz squashed and added a comment regarding concurrency for parsedClasses

**Review by lrytz — APPROVED** [review]

**Pull request merged**